### ISSUE TITLE
Rotated TTF text not rendered correctly using raqm

### ISSUE THREAD

**Issue by dwhile** [libgd/libgd#790]
**Describe the bug**
Using PHP to produce an image, when the text is rotated 90 degrees (i.e. vertical) the text is rendered with the characters on top of one another.

This only appears to be a problem with True Type fonts.

**To Reproduce**
```
<?php
header('Content-Type: image/png');
$im = imagecreatetruecolor(30, 300);
$white = imagecolorallocate($im, 255, 255, 255);
$grey = imagecolorallocate($im, 128, 128, 128);
$black = imagecolorallocate($im, 0, 0, 0);
imagefilledrectangle($im, 0, 0, 29, 299, $white);
$text = 'Testing...';
$font = '/usr/share/fonts/msttcorefonts/arial.ttf';
imagettftext($im, 20, 90, 26, 221, $grey, $font, $text);
imagettftext($im, 20, 90, 25, 220, $black, $font, $text);
imagepng($im);
?>
```

Using this code produces the problem.

**Expected behavior**
A white box with "Testing....." displayed vertically read from bottom to top.

**Actual results**
A white box with the characters on top of one another. See https://www.while.org.uk/try.png

**Environment (please complete the following information):**
 - OS: Fedora Linux
 - Version 35
- GD V2.3.2 (also tried with GD2.3.3)
- Freetype 2.11.0
**Additional context**
This only appears to be a recent problem when upgrading from Fedora 34 to Fedora 35. I initially reported it to the PHP development team who said it was an upstream problem and to report it here.

**Comment by cmb69** [comment]
> This only appears to be a recent problem when upgrading from Fedora 34 to Fedora 35.

@remicollet, already heard about that issue?

**Comment by remicollet** [comment]
I confirm I can reproduce this issue on freshly released F35

As same test is OK using same version of libgd (2.3.2) and with different PHP versions (7.4, 8.0, 8.1) on Fedora 33 and 34, I don't think this is an issue in libgd itself, but rather in some used libraries (ex: freetype 2.10 vs 2.11). Need more investigation

**Comment by dwhile** [comment]
Ok I installed a fresh copy of Fedora 34 (using a VM) and confirmed that the output was correct (text was vertical and spaced correctly). I then upgraded freetype to version 2.11.0 and reran the test. The output was still correct. so the problem doesn't appear to be with freetype 2.11.0

**Comment by dwhile** [comment]
Ok - by upgrading each library in turn and checking the result it appears the problem lies in libraqm. Upgrading that library caused the problem. Downgrading it again restores the correct output.

**Comment by dwhile** [comment]
Just tried on a Fedora 35 release - downgrade the libraqm library and everything works normally.
Interestingly the version numbers of the library on F34 and F35 appear to be the same.

**Comment by remicollet** [comment]
> Just tried on a Fedora 35 release - downgrade the libraqm library and everything works normally. Interestingly the version numbers of the library on F34 and F35 appear to be the same.

downgrade to which version ? (0.7.0-9.fc35 or 0.7.2-1.fc34 or other ?)

**Comment by dwhile** [comment]
It downgrades to 0.7.0-9.fc35

**Comment by remicollet** [comment]
For memory, [#2022957](https://bugzilla.redhat.com/show_bug.cgi?id=2022957)

**Comment by remicollet** [comment]
It works properly with libraqm-0.7.2-1.fc34 on F35
So this looks like a build chain issue... very hard to diagnose...

**Comment by pierrejoye** [comment]
Talking about libraqm, it seems to me to be rarely used by other projects. 

What do you think about using something more common like harfbuzz? Especially in 3.0 as we can just render with GD as well without using Cairo.

**Comment by pierrejoye** [comment]
@cmb69 harfbuzz supports  LTR/TTB/RTL/BTT nicely. It is a bit more codes as harfbuzz fits one layer lower (basically one an editor would use and then handles the rendering itself).

Alternatively, recent Pango also supports custom raster, so it could be done with GD as well. 

The reason I asked is when I worked on the CI, I looked into RAMQ implementation, the lib and how it is done in GD. I saw a lot of duplication and may need a rewrite to be clean. Mainly to avoid introducing bug again by changing non RAMQ part and leave the RAMQ part alone. Also RAMQ is not yet widely available on all distros (as of 2 months ago).

**Comment by remicollet** [comment]
FYI I chose to disable libraqm usage on Fedora >= 35

@dwhile please see https://bodhi.fedoraproject.org/updates/FEDORA-2021-53262e3123

**Comment by remicollet** [comment]
Culprit found, reported as https://github.com/HOST-Oman/libraqm/issues/141

**Comment by pierrejoye** [comment]
not a gd issue, closing it.

**Comment by khaledhosny** [comment]
I think it is a gd issue. Raqm does not have a concept of rotated text. The related change is applying FreeType transformation to get correct glyph metrics, but Raqm itself does not set the transformation.

**Comment by khaledhosny** [comment]
I suspect the application of `cos_a` and `sin_a` to the glyph positions to be the culprit, since Raqm is applying the transformation to the positions it is returning as well (something it should have done from the start, but FreeType lacked API to get the transformation): https://github.com/libgd/libgd/blob/d6baf4c5c97726881360e6d01307441066bc9425/src/gdft.c#L1535-L1536

**Comment by cmb69** [comment]
@khaledhosny, thanks! So we would need to detect whether libraqm >= 0.7.2 and whether libfreetype >= 2.11. Is there maybe some simpler way to accomplish that?

**Comment by khaledhosny** [comment]
That is probably the simplest way. Or You can check for if FreeType has `FT_Get_Transform()`, but Raqm does not have some equivalent function to check for.

**Comment by pierrejoye** [comment]
FT get transform exists since years, it is part already of the min freetype version we require.

I would rather increase the min required version of raqm. rather than doing build or runtime error.

**Comment by pierrejoye** [comment]
Can we close that one as it seems to be a libqarm issue?

**Comment by cmb69** [comment]
I think we need to do something. The issue happens with libraqm >= 0.7.2, so requiring that version wouldn't solve the problem (and we still need to cater to builds without libraqm).

**Comment by pierrejoye** [comment]
is it a bug in libqarm right?

as of the build, what do you mean?

**Comment by cmb69** [comment]
> is it a bug in libqarm right?

See https://github.com/libgd/libgd/issues/790#issuecomment-982894815. So libraqm changed its behavior, and we somehow have to deal with it.

**Comment by pierrejoye** [comment]
thanks :)

not sure to understand this comment.

So raqm applies transformation ton the glyphs now but didn't do it before? or do I read it wrong?

**Comment by pierrejoye** [comment]
I will need to test and read libqram changes.

It relies on the above mentioned libraires too btw.

**Comment by cmb69** [comment]
> So raqm applies transformation ton the glyphs now but didn't do it before? or do I read it wrong?

That's how I understand it. :)

**Comment by pierrejoye** [comment]
Using Master, with:

```
/**
 * Bug 790, Rotated TTF text not rendered correctly using RAQM
 *
 * RAQM 0.7.2+ BC Break
 *
 * See <https://github.com/libgd/libgd/issues/790>
 */

#include "gd.h"
#include "gdtest.h"

int main()
{
    gdImagePtr im = gdImageCreate(100, 100);

    int rect[8];
    int bg = gdImageColorAllocate(im, 255, 255, 255);
    int fg = gdImageColorAllocate(im, 0, 0, 0);
    char *path = gdTestFilePath("freetype/DejaVuSans.ttf");
    char *res = gdImageStringFT(im, rect, fg, path, 12, M_PI/2, 10, 100, "ABCDEF");
    FILE *fp;

    gdTestAssert(res == NULL);
    free(path);

    fp = fopen("out.png", "wb");
    gdImagePng(im, fp);
	fclose(fp);

    gdImageDestroy(im);
    return gdNumFailures();
}
```
I get 
with RAQM master:
![out](https://user-images.githubusercontent.com/282408/150664250-c5af20e7-9586-4e8a-b61b-20c6603c48a7.png)

Without RAQM:
![out](https://user-images.githubusercontent.com/282408/150664287-b7f2545e-9310-4753-a763-fc6428884f22.png)

So I am really not sure what to do at this point :)

**Comment by pierrejoye** [comment]
Except if there are something new or someone can find cases where it fails, I will close this bug next week.

**Comment by khaledhosny** [comment]
[`FT_Get_Transform()`](https://freetype.org/freetype2/docs/reference/ft2-base_interface.html#ft_get_transform) was introduced in FreeType 2.11, Raqm needs it to get the transform applied to the `FT_Face` and the new behavior depends on this (the transform was not previously applied because FreeType provided no way to get the transform, `FT_Get_Transform()` was added precisely for this).

**Comment by pierrejoye** [comment]
> [`FT_Get_Transform()`](https://freetype.org/freetype2/docs/reference/ft2-base_interface.html#ft_get_transform) was introduced in FreeType 2.11, Raqm needs it to get the transform applied to the `FT_Face` and the new behavior depends on this function’s presence (Raqm did not previously apply the transform because FreeType provided no way to get it, `FT_Get_Transform()` was added precisely for this).

Ah right. I was confused with another API.

So the bug fixes is a kind of a BC break:

```C
#ifdef HAVE_FT_GET_TRANSFORM

    {
      FT_Matrix matrix;
      hb_glyph_position_t *pos;
      unsigned int len;

      FT_Get_Transform (hb_ft_font_get_face (run->font), &matrix, NULL);
      pos = hb_buffer_get_glyph_positions (run->buffer, &len);
      for (unsigned int i = 0; i < len; i++)
      {
        _raqm_ft_transform (&pos[i].x_advance, &pos[i].y_advance, matrix);
        _raqm_ft_transform (&pos[i].x_offset, &pos[i].y_offset, matrix);
      }
    }
#endif
```

As in, as you mention, no transformation at all were applied before.

That's a rather big change, with quite some cases to handle now. As raqm supports both FT, with or without FT_Get_Transform, and the same versions will behave completely differently depending on the Freetype version used. I am not sure it is something we can handle at this point. I cannot even imagine the support pain increase for end users (php and co) for it.

I tend to think to go with allowing RAQM only if FT 2.11 is available (also I am not sure how to check RAQM is built against it either so that won't solve everything).

By the way,@khaledosny which versioning scheme do you follow? It could help us to handle such situations.

**Comment by pierrejoye** [comment]
One option I would suggest is:

raqm_layout_with_options(raqm_t *rq, raqm_options_t options)

where options could contain something like transform [true|false]

That way BC can be kept easily.

Another question come to my mind, any reason why there is no _manual_ transformation in case of FT_Get_Transform does not exist?

**Comment by pierrejoye** [comment]
@cmb69 @remicollet I tend to do:

- Disable RAQM if FT < 2.11 altogether even if older version are present (see comment above)
- Mark RAQM support as experimental until 1.0.0 is reached

I do like it, nice API and al, very good work. However this kind of changes in minor/patch releases being considered as bug fixes is a no go from our side (support++).

thoughts?

**Comment by cmb69** [comment]
Well, < 1.0.0 there is no stability guarantee. I'm fine with marking RAQM support as experimental for the time being.

**Comment by khaledhosny** [comment]
We didn’t anticipate this Raqm change to be a backward compatibility issue, but this ship has sailed. I think checking for Raqm version and the presence of `FT_Get_Transform()` should be good enough.

**Comment by pierrejoye** [comment]
PR added implementing the above suggestion.

**Comment by pierrejoye** [comment]
> We didn’t anticipate this Raqm change to be a backward compatibility issue, but this ship has sailed. I think checking for Raqm version and the presence of `FT_Get_Transform()` should be good enough.

Yeah, testing for 2.11+ or disable RAQM.

Thanks for the clarification about BC policies, experimental makes perfect sense for libGD then, thanks! :)

**Comment by khaledhosny** [comment]
I’m not sure how the PR mentioned above fix the issue, since libgd is still applying the transformation. Anyway, I was already checking libgd code and this patch fixes the positioning of rotated text (needs to be adapted for disabled Raqm case):
```diff
diff --git a/src/gdft.c b/src/gdft.c
index acffdeea..edb7ba67 100644
--- a/src/gdft.c
+++ b/src/gdft.c
@@ -436,6 +436,7 @@ gdTcl_UtfToUniChar (const char *str, Tcl_UniChar * chPtr)
 typedef struct {
        unsigned int index;
        FT_Pos x_advance;
+       FT_Pos y_advance;
        FT_Pos x_offset;
        FT_Pos y_offset;
        uint32_t cluster;
@@ -483,6 +484,7 @@ textLayout(uint32_t *text, int len,
                info[i].x_offset = glyphs[i].x_offset;
                info[i].y_offset = glyphs[i].y_offset;
                info[i].x_advance = glyphs[i].x_advance;
+               info[i].y_advance = glyphs[i].y_advance;
                info[i].cluster = glyphs[i].cluster;
        }

@@ -520,6 +522,7 @@ textLayout(uint32_t *text, int len,
                if (delta.x != 0)
                        info[count - 1].x_advance += delta.x;
                info[count].x_advance = face->glyph->metrics.horiAdvance;
+               info[count].y_advance = 0;
                info[count].cluster = count;

                /* carriage returns or newlines */
@@ -1131,7 +1134,7 @@ BGD_DECLARE(char *) gdImageStringFTEx (gdImagePtr im, int *brect, int fg, const
         */
        gdCache_head_t *tc_cache;
        /* Tuneable horizontal and vertical resolution in dots per inch */
-       int hdpi, vdpi, horiAdvance, xshow_alloc = 0, xshow_pos = 0;
+       int hdpi, vdpi, horiAdvance, vertAdvance, xshow_alloc = 0, xshow_pos = 0;
        FT_Size platform_specific, platform_independent;

        if (strex) {
@@ -1462,6 +1465,7 @@ BGD_DECLARE(char *) gdImageStringFTEx (gdImagePtr im, int *brect, int fg, const
                }

                horiAdvance = info[i].x_advance;
+               vertAdvance = info[i].y_advance;

                if (brect) {
                        /* only if need brect */
@@ -1531,15 +1535,22 @@ BGD_DECLARE(char *) gdImageStringFTEx (gdImagePtr im, int *brect, int fg, const
                        (the estimate was rounded up to next 1/METRIC_RES, so this should fit) */
                        FT_Pos pen_x = penf.x + info[i].x_offset;
                        FT_Pos pen_y = penf.y - info[i].y_offset;
+#if 0
                        gdft_draw_bitmap (tc_cache, im, fg, bm->bitmap,
                                          (int)(x + (pen_x * cos_a + pen_y * sin_a)*hdpi/(METRIC_RES*64) + bm->left),
                                          (int)(y - (pen_x * sin_a - pen_y * cos_a)*vdpi/(METRIC_RES*64) - bm->top));
+#else
+                       gdft_draw_bitmap (tc_cache, im, fg, bm->bitmap,
+                                         (int)(x + pen_x*hdpi/(METRIC_RES*64) + bm->left),
+                                         (int)(y - pen_y*vdpi/(METRIC_RES*64) - bm->top));
+#endif

                        FT_Done_Glyph (image);
                }


                penf.x += horiAdvance;
+               penf.y += vertAdvance;
        }

        gdFree(text);
```

Apparently libgd was ignoring the `y_advance`, so even if it stopped applying the transformation the text would still be badly positioned when rotated.

**Comment by cmb69** [comment]
Thanks, @khaledhosny! It might also make sense to check how that's handled by PHP's bundled libgd (there are certainly differences).

**Comment by pierrejoye** [comment]
Thanks, let me check the availability of y_advance (version).

My PR is only about disabling raqm support if ft 2.11+ is not used.

**Comment by pierrejoye** [comment]
@khaledhosny also I wonder what you plan to do for that function. Having completely different behavior based on ft version is not that good. 

It obviously needs get transform, it should fail to build of not present no? I can send a PR if desired

**Comment by khaledhosny** [comment]
FreeType 2.11 was released only 6 months ago, requiring it would be a bit too much given that not all Raqm users transformations on FT_Face (so far this is the only case I’m aware of).

**Comment by pierrejoye** [comment]
@khaledhosny I understand :)

However this function promise is to apply transformation if required.

That promise is very hard to determine or keep given its implementation.

If there are cases for both with or without transform, it could be better to have two functions or using options. And it should either fails at build, or gracefully at runtime so the calling code can handle the error.

As it is now, it is a lucky game. Anyone updating will suddenly see transformation, or the other way around, anyone using it with ft 2.11 but having users with 2.10 will have no transformation.

Long story short, either it has to and can apply transformation or it does not. Not a mix of both depending on external factors with no way to see programmatically that none was applied. :)

**Comment by pierrejoye** [comment]
About y_advance, it is from ramq and used by ramq only. I will need to test it and do a PR with tests, thanks :)

I first thought by reading it that it was something from FT (which I could not remember at all but vert advance for vertical layout only).

**Comment by khaledhosny** [comment]
Any new change regarding transforms in Raqm is going to make the situation even worse. At this point I think you can either:
1. Require FreeType 2.11 when Raqm is enabled and update the code to assume Raqm always applies the transformation.
2. Check FreeType and Raqm versions and apply the transformation only conditionally.

**Comment by remicollet** [comment]
> @cmb69 @remicollet I tend to do:
> 
>     * Disable RAQM if FT < 2.11 altogether even if older version are present (see comment above)
> 
>     * Mark RAQM support as experimental until 1.0.0 is reached
> 
> 
> I do like it, nice API and al, very good work. However this kind of changes in minor/patch releases being considered as bug fixes is a no go from our side (support++).
> 
> thoughts?

I'm fine for dropping support for RAQM, it is what I've done for the downstream packaging of libgd in Fedora

**Comment by pierrejoye** [comment]
@khaledhosny will go for disabling it for now, with ft  <2.11

about that function, raqm is not yet 1.0, that is the time to make apis rock solid. This randomly defined behavior is not what I may look for in a rock solid API :)

**Comment by khaledhosny** [comment]
Disabling it without adapting the code is pointless, people building with newer FreeType will still get bad rendering of rotated text.

Ram API and ABI is stable, this unintentional change in behavior doesn't directly affect this stability. I might actually go ahead and require FreeType 6.12, it now seems the less complicated option, and lings can then check for Ram version and adapt the code accordingly.

I wouldn't recommend flat out building without Raqm, rotated text us fringe use case IMHO, while Raqm-less libgd can’t display text for languages used by billions of people.

**Comment by pierrejoye** [comment]
Hi Khaled,


It supports many languages, including ltr ones.

Text handling, as you know very well, is a complex task :)

We also have gd-pango.

My only point now is to avoid different random behaviors and support
increase. And that's what this 1st step does.

To do the clean way is a much bigger task as it needs quite some
refactoring (the initial raqm addition was not so good, many code
duplications). We will get there.

Only to be a bit more clear, APIs predictable behaviors and stability are
key. Hence why I insist a little bit on that one. Whether there is a
majority of users not using rotation is not really relevant imho.

Requiring 2.12 sounds like a plan, with the down side that it is really not
widely available. Again, it is relevant for transformed glyph only.

Long run, it will pay off tho'.

**Comment by khaledhosny** [comment]
FYI https://github.com/HOST-Oman/libraqm/pull/151

**Comment by pierrejoye** [comment]
good morning

excellent, thank you very much :)

once released, I will use that version as a minimum version, updating the
PR accordingly and refactor the raqm support as well.

**Comment by khaledhosny** [comment]
https://github.com/HOST-Oman/libraqm/releases/tag/v0.9.0
